src/librustc/traits/error_reporting.rs

```diff
@@ -817,6 +818,58 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
         }
     }
 
+    /// Whenever references are used by mistake, like `for (i, e) in &vec.iter().enumerate()`,
+    /// suggest removing these references until we reach a type that implements the trait.
+    fn suggest_remove_reference(&self,
+                                obligation: &PredicateObligation<'tcx>,
+                                err: &mut DiagnosticBuilder<'tcx>,
+                                trait_ref: &ty::Binder<ty::TraitRef<'tcx>>) {
+        let ty::Binder(trait_ref) = trait_ref;
+        let span = obligation.cause.span;
+
+        if let Ok(snippet) = self.tcx.sess.codemap().span_to_snippet(span) {
+            let refs_number = snippet.chars()
+                .filter(|c| !c.is_whitespace())
+                .take_while(|c| *c == '&')
+                .count();
+
+            let mut trait_type = trait_ref.self_ty();
+            let mut selcx = SelectionContext::new(self);
+
+            for refs_remaining in 0..refs_number {
+                if let ty::TypeVariants::TyRef(_, ty::TypeAndMut{ ty: t_type, mutbl: _ }) =
+                    trait_type.sty {
+
+                    trait_type = t_type;
+
+                    let substs = self.tcx.mk_substs_trait(trait_type, &[]);
+                    let new_trait_ref = ty::TraitRef::new(trait_ref.def_id, substs);
+                    let new_obligation = Obligation::new(ObligationCause::dummy(),
+                                                         obligation.param_env,
+                                                         new_trait_ref.to_predicate());
+
+                    if selcx.evaluate_obligation(&new_obligation) {
+                        let remove_refs = refs_remaining + 1;
+
+                        let suggest_snippet = snippet.chars()
+                            .filter(|c| !c.is_whitespace())
+                            .skip(remove_refs)
+                            .collect::<String>();
```

Comment:
Wouldn't this turn something like `& ( a, b )` into `(a,b)` or `& & & & v   .iter()    .enumerate()` into `v.iter().enumerate()`?

Reply:
So, for things like tuples, arrays and slices, we would like to keep the current formatting. For things like very long chains of method calls split each on its own line (because having them all in the same line would be over 100 chars wide) we would also like to keep the formatting.

I think you should instead advance the iterator while the char is whitespace or `&`, and return the everything after that. That way you trim the beginning the way you want, but keep the rest of the code with the current formatting.

Reply:
Agreed.